### PR TITLE
Support union types when getting value

### PR DESCRIPTION
**Pull request by huskicsuad**
☝️

### CODE REVIEW

File: tests/DeSerializationTest.php [hunk 1]
```diff
@@ -600,4 +601,25 @@ public function testRequiredPropertyMissing()
...
+            'value' => 'website-name'
+        ]);
+
+        $this->assertTrue(is_int($caInt->value));
+        $this->assertEquals(5, $caInt->value);
+
+        $this->assertTrue(is_string($caString->value));
+        $this->assertEquals('website-name', $caString->value);
+
+
```
Comment (huskicsuad) [inline tests/DeSerializationTest.php:623]:
```suggestion
```

**Review by khabarov-a — APPROVED** [review]

File: tests/Definitions/CustomAttribute.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class CustomAttribute
+{
+    use JsonSerialize;
+
+    #[Json]
+    public string $key;
+
+    #[Json]
+    public int|string $value;
```
Comment (khepin) [inline tests/Definitions/CustomAttribute.php:16]: We need some tests with `int|string|null` 

We also need to decide what to do if the type is `Class1|Class2`

File: src/Json.php [hunk 3]
```diff
@@ -70,6 +71,20 @@ public function retrieveValue(?array $data, ?ReflectionNamedType $type = null)
             $data = $data[$pathBit];
         }
 
+        if ($type instanceof ReflectionUnionType) {
+            $typesByType = $type->getTypes();
+            foreach ($type->getTypes() as $type) {
+                $typesByType[$type->getName()] = $type;
+            }
+            $type = match (gettype($data)) {
```
Comment (khepin) [inline src/Json.php:79]: If we're going to overwrite `$type`, this needs to be done after the checks for `$type->allowsNull()`.

If the type is `string|int|null` for example, the `ReflectionUnionType` will have `->allowsNull()` true, but the `string` one will not.

File: src/Json.php [hunk 4]
```diff
@@ -70,6 +71,47 @@ public function retrieveValue(?array $data, ?ReflectionNamedType $type = null)
             $data = $data[$pathBit];
         }
 
+        if ($type instanceof ReflectionUnionType) {
```
Comment (huskicsuad) [inline src/Json.php:74]: @khepin can you give this another look? couple of new things
 - support null within union type
 - support array OR custom object within union type, not both at the same time because the input structure will look the same. 
 - support a single custom object within union type because the input structure will look the same.

**Comment by huskicsuad** [comment]
closing in favor of: https://github.com/square/pjson/pull/31
